cnd/src/network/comit_ln.rs

```diff
@@ -577,3 +577,281 @@ impl NetworkBehaviourEventProcess<oneshot_behaviour::OutEvent<finalize::Message>
+
+        // let alice_swarm_arc = Arc::new(alice_swarm);
+        // let bob_swarm_arc = Arc::new(bob_swarm);
+
+        let alice_swarm = Arc::new(Mutex::new(alice_swarm));
+        let bob_swarm = Arc::new(Mutex::new(bob_swarm));
+
+        // construct future to asynchroniously kick off the swap communication for alice
+        let future_init = start_communication(
+            alice_swarm.clone(),
+            bob_swarm.clone(),
+            swap_params_alice,
+            swap_params_bob,
+            NodeLocalSwapId::default(),
+            NodeLocalSwapId::default(),
+        );
+
+        // future to poll for the finalized event of alice behaviour
+        let alice_future = async {
+            println!("Start listening for events for alice");
+            loop {
+                // sleep(time::Duration::from_millis(300));
+
+                let mut guard = alice_swarm.lock().await;
+
+                let next = guard.next_event().await;
+
+                println!("Got alice event: {:?}", next);
+                match next {
+                    SwarmEvent::Behaviour(behavior_event) => match behavior_event {
+                        BehaviourOutEvent::SwapFinalized {
+                            local_swap_id,
+                            swap_params,
+                            secret_hash,
+                            ethereum_identity,
+                        } => {
+                            assert_eq!(ethereum_identity_bob, ethereum_identity);
+                            break;
+                        }
+                    },
+                    _ => {
+                        continue;
+                    }
+                }
+            }
+        };
+
+        // future to poll for the finalized event of bob behaviour
+        let bob_future = async {
+            println!("Start listening for events for bob");
+            loop {
+                // sleep(time::Duration::from_millis(300));
+
+                let mut guard = bob_swarm.lock().await;
+                let next = guard.next_event().await;
+
+                println!("Got bob event: {:?}", next);
+                match next {
+                    SwarmEvent::Behaviour(behavior_event) => match behavior_event {
+                        BehaviourOutEvent::SwapFinalized {
+                            local_swap_id,
+                            swap_params,
+                            secret_hash,
+                            ethereum_identity,
+                        } => {
+                            assert_eq!(ethereum_identity_alice, ethereum_identity);
+                            break;
+                        }
+                    },
+                    _ => {
+                        continue;
+                    }
+                }
+            }
+        };
+
+        // join all futures
+        let joined = join3(alice_future, bob_future, future_init);
+
+        runtime.block_on(joined);
```

Comment:
Why do you join the init future into this?  Why not just run it first then join the other two?